Sprint 8 solution time and duration

kafi40 opened the pull request:
В данном ТЗ добавлены дата создания и окончания задачи. Изменена работа методов с использованием функционального программирования. Улучшено отображение задач в консоле с использованием полученных знаний. Добавлены новые тесты.

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли

kafi40 committed:
Добавить сохранение задач
Добавить сохранение истории просмотра

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли
Добавить сортировку задач по времени
Добавить новые тесты

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли
Добавить сортировку задач по времени
Добавить новые тесты

kvach2avl commented on src/task/Task.java:
```diff
@@ -49,6 +60,21 @@ public void setTaskStatus(TaskStatus taskStatus) {
+    }
+
+    public LocalDateTime getEndTime() {
+        if (startTime != null) {
```
из-за того, что duration тоже может быть null-ом, лучше немного расширить условие: if (startTime != null && duration != null) ...

kvach2avl commented on src/task/EpicTask.java:
```diff
@@ -1,11 +1,15 @@
 public class EpicTask extends Task {
-
     private ArrayList<SubTask> subTasks;
+    private LocalDateTime endTime;
```
Зачем иметь отдельную переменную для окончания эпика, если это время каждый раз считается заново в методе getEndTime() ?
Либо эта переменная не нужна, либо (что лучше) надо вычислить время начала/окончания и длительность в менеджере (в момент создания/редактирования/удаления подзадач и эпиков), запомнить их и пользоваться.

kvach2avl commented on src/task/EpicTask.java:
```diff
@@ -24,4 +28,21 @@ public void removeSubTask() {
+        this.setStartTime(optionalStartTime.map(Task::getStartTime).orElse(null));
+        return optionalStartTime.map(Task::getStartTime).orElse(null);
+    }
+
```
По ТЗ длительность эпика должна вычисляться, но ни здесь ( по аналогии с getStartTime()/getEndTime() ), ни в менеджере такого кода не нашел.
Надо добавить вычисление длительности эпика (найти сумму длительностей всех его подзадач)

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -9,18 +9,35 @@
+    private final TreeSet<Task> taskTreeSet;
+    private boolean isSorted = false;
     public static int taskIdCounter = 0;
     private final TaskHistory taskHistory = new InMemoryTaskHistory();
```
В ТЗ предыдущих спринтов декларировалась необходимость создать утилитный класс Managers со вспомогательными методами, в частности:
    public static HistoryManager getDefaultHistory() { ... }
как раз с такой целью, чтобы удобнее было создавать менеджера истории.

В идеале здесь должно быть что такое:
    private final TaskHistory taskHistory = Managers.getDefaultHistory();

kvach2avl commented on src/interfaces/TaskManager.java:
```diff
@@ -1,15 +1,17 @@
 import java.io.IOException;
+import java.util.TreeSet;
 
 public interface TaskManager {
```
Может ли такое быть, что в PR прислана не совсем та работа, которая планировалась?

Смущает следующее: интерфейс менеджера весьма сильно не соответствует ТЗ предыдущих спринтов.

Крупными мазками:
1. никакого интерактива с пользователем не требуется. Все задачи, подзадачи, эпики должны создаваться в коде (в тестовом, в Main, но НЕ посредством взаимодействия через консоль);
2. методы нахождения чего-либо по идентификатору (которые сейчас реализованы в findTaskById) должны принимать в метод аргументом идентификатор, возвращать из метода ссылку на объект. Никаких консолей, никаких файлов;
3. класс FileBackedTaskManager должен расширять InMemoryTaskManager , перекрывать в нём методы создания, обновления и удаления с целью внедрения в них метода save(). Метод save() не содержит никаких аргументов и сохраняет текущее состояние менеджера задач и менеджера истории в файл. Из подобного файла класс FileBackedTaskManager должен уметь восстанавливать собственное состояние. Т.е. метод FileBackedTaskManager.getData(), реализованный сейчас, должен не список прочитанных задач возвращать, а экземпляр менеджера, в котором присутствуют те самые задача, подзадачи и эпики, которые были сохранены в файле.
4. Вот примерная структура интерфейса, какой она должна быть:
```
public interface TaskManager {
	List<Task> getTasks();
	List<SubTask> getSubtasks();
	List<EpicTask> getEpics();
	List<SubTask> getEpicSubtasks(int epicId);
	Task getTask(int id);
	SubTask getSubtask(int id);
	EpicTask getEpic(int id);
	int addNewTask(Task task);
	int addNewEpic(EpicTask epic);
	int addNewSubtask(SubTask subtask);
	void updateTask(Task task);
	void updateEpic(EpicTask epic);
	void updateSubtask(SubTask subtask);
	void deleteTask(int id);
	void deleteEpic(int id);
	void deleteSubtask(int id);
	void deleteTasks();
	void deleteEpics();
	void deleteSubtasks();
	List<Task> getHistory();
	List<Task> getPrioritizedTasks();
}
```

Соответствие присланной работы требованиям _этого_ ТЗ даже не проверял, поскольку она не соответствует как минимум трём предыдущим.

kvach2avl reviewed (commented):
Есть вопросы к реализации требований ТЗ. Подробности ниже

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+                    .filter(task -> task.getStartTime() != null)
+                    .collect(Collectors.toSet());
+            isSorted = true;
+            taskTreeSet.addAll(taskSet);
```
Этот метод будет только добавлять данные и никогда не удалит старые, что некорректно

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+    @Override
+    public TreeSet<Task> getPrioritizedTasks() {
+        if (!isSorted) {
+            Set<Task> taskSet = allTypeTasks.values().stream()
```
в нынешней реализации менеджера в allTypeTasks попадают все задачи, тогда как getPrioritizedTasks должен возвращать только простые задачи и подзадачи. Поэтому корректнее организовать его как-то так:
```
    public Collection<Task> getPrioritizedTasks() {
        if (!isSorted) {
            taskTreeSet.clear();

            List<Task> candidates = new ArrayList<>(tasks.values());
            candidates.addAll(subTasks.values());

            List<Task> taskSet = candidates.stream()
                    .filter(t -> t.getStartTime() != null)
                    .collect(Collectors.toList());

            taskTreeSet.addAll(taskSet);

            isSorted = true;
        }

        return taskTreeSet;
    }
```

За идею проверять саму необходимость повторной сортировки (флаг isSorted) - зачёт!

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+                            - Math.max(startTime.atZone(ZONE_ID).toInstant().toEpochMilli(),
+                            t.getStartTime().atZone(ZONE_ID).toInstant().toEpochMilli());
+                    if (result > 0) {
+                        System.out.println("На это время уже есть задача");
```
Рекомендация на будущее: если обнаружена какая-то ситуация, нарушающая ограничения бизнес-логики (в данном случае, это конфликт времен исполнения), то следует не жаловаться в консоль, а бросить исключение (наследника непроверяемого исключения, например, собственный TaskValidationException).
В данном случае, поскольку приложение получилось интерактивным, можно оставить так.

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+                startTime = optionalStartTime.get();
+                duration = Duration.ofMinutes(optionalDuration.get());
+                for (Task t : taskTreeSet) {
+                    long result = Math.min(startTime.plus(duration).atZone(ZONE_ID).toInstant().toEpochMilli(),
```
Ни startTime, ни duration внутри цикла не меняются, поэтому всю арифметику с ними можно вынести вне цикла в отдельные переменные:
```
                long sEnd = startTime.plus(duration).atZone(ZONE_ID).toInstant().toEpochMilli();
                long sStart = startTime.atZone(ZONE_ID).toInstant().toEpochMilli();
```
тогда код внутри цикла сразу станет проще для восприятия:
```
                    long result = Math.min(sEnd, t.getEndTime().atZone(ZONE_ID).toInstant().toEpochMilli())
                                - Math.max(sStart, t.getStartTime().atZone(ZONE_ID).toInstant().toEpochMilli());
```

Вторая рекомендация касается работы с временной зоной - поскольку все времена задаются в одной зоне, то лучше не приводить их к миллисекундам, а работать с ними как с временами:
```
if (!startTime.isBefore(t.getStartTime()) && !startTime.isAfter(t.getEndTime())) ... // случай, когда начало задачи попадает на период исполнения другой задачи
```

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+            if (optionalDuration.isPresent()) {
+                startTime = optionalStartTime.get();
+                duration = Duration.ofMinutes(optionalDuration.get());
+                for (Task t : taskTreeSet) {
```
Из цикла проверки следует исключить задачу, которая _уже_ присутствует в менеджере, но которую решили обновить ( из editTaskById() ). В противном случае она будут проверена сама с собой и выдаст "самопересечение".
Проверить достаточно по id: if (t.getTaskId().equals(task.getTaskId())) continue;

kvach2avl commented on src/memory/InMemoryTaskManager.java:
```diff
@@ -340,35 +364,116 @@ public  void getHistory() {
+                            t.getEndTime().atZone(ZONE_ID).toInstant().toEpochMilli())
+                            - Math.max(startTime.atZone(ZONE_ID).toInstant().toEpochMilli(),
+                            t.getStartTime().atZone(ZONE_ID).toInstant().toEpochMilli());
+                    if (result > 0) {
```
Если result == 0, это тоже не очень хорошо. Это значит, что начало какой-то задачи совпадает с окончанием другой

kvach2avl reviewed (commented):
часть II

kafi40 committed:
Добавить начало и конец выполнения задачи
Изменить логику методов с применением лямбда выражений
Улучшить отображение задач в консоли
Добавить сортировку задач по времени
Добавить новые тесты

kafi40 committed:
Исправить замечания по коду

kvach2avl commented on src/task/EpicTask.java:
```diff
@@ -24,4 +28,21 @@ public void removeSubTask() {
         subTasks.clear();
     }
 
```
Замечание в части длительности эпика остаётся в силе. По ТЗ - цитирую: Продолжительность эпика — сумма продолжительностей всех его подзадач.
Необходимо реализовать это требование ТЗ. В имеющейся концепции это можно сделать аналогично реализации getStartTime() - пробежаться стримом по всем подзадачам, выбрать все с ненулевым duration и вернуть их сумму.

kvach2avl reviewed (commented):
Принято. Замечание в части длительности эпика просьба не забыть.
